### ISSUE TITLE
isomorphic example application / server side rendering

### ISSUE THREAD

**Issue by srghma** [slamdata/purescript-halogen#610]
is there any example application with isomorphic / server side rendering? 

e.g. Pux framework has one https://github.com/alexmingoia/pux-starter-app/tree/isomorphic

**Comment by garyb** [comment]
No, there's nothing like that. There is a library that provides string rendering for the Halogen VDOM (https://github.com/slamdata/purescript-halogen-vdom-string-renderer), but it's only had very light use.

**Comment by srghma** [comment]
@garyb tnx

this allows to render generated VDom to string, but how to wait for application to execute initial requests and return state?

but how to have something like nextjs:
1. run halogen application code on server
2. wait for application to execute first request (e.g. nextjs has a [`getServerSideProps`/`getInitialProps`](https://nextjs.org/docs/api-reference/data-fetching/getInitialProps#recommended-use-getstaticprops-or-getserversideprops-instead) method, pux has a [`waitState`](https://github.com/alexmingoia/pux-starter-app/blob/f29ac097dbfca9ac36805e1b50ce4568afc2fb0e/src/Server.purs#L58) method)
3. extract state and render app vdom to string

```
const initialState = {}
const updatedApp = execGetServerSideProps(app, initialState)
const state = extractState(updatedApp)
```

4. send `state` as json and `updatedApp` as string
5. on client: first show html, then replace html with mounted application

```
documentFind('body#app').html = prerenderedHtml
mountApp('body#app', app, state)
```

**Comment by garyb** [comment]
That's definitely a tricky question in Halogen, given that it is designed explicitly to support situations where not all the state is recorded in the component states (for embedding 3rd party components and such where they maintain their own state outside of Halogen, and are interacted with through queries instead). I guess you'd just have to not use any components that do that if using server side rendering.

There are plenty of options for getting the state out otherwise:

1. Architect the app so the top level component has all the app state in it, using inputs and messages to pass the state down into child components / notify changes, etc.
2. Similar to 1, just make the whole app one component and uses lenses and stuff for managing state composition.
4. Give all components a query that returns their serialised state, and then have it so that calling the query on the root component calls the query on its children, and so on.
3. Use `ReaderT + Aff` and have component write their serialised states into a `Ref`.
4. Make an alternate driver that can expose the component states. Would need an extra constraint on the components to ensure their states can be serialised or something though.

etc.

There isn't really a set way of doing things in Halogen since it's not meant to be a dictatorial framework. There are a lot of common patterns people use, but I wouldn't want to change things in a way that forces people to do things one particular way, so anything like this is probably going to have to be an ad-hoc thing built on top of what Halogen already provides, rather than something it'll  support as part of the library itself.

**Comment by srghma** [comment]
The hydrate method is not implemented for ssr rendering

https://reactjs.org/docs/react-dom.html#hydrate

https://github.com/snabbdom/snabbdom/blob/master/src/tovnode.ts

**Comment by srghma** [comment]
Done https://github.com/srghma/purescript-halogen-nextjs
